File: code/ex918.java
class Solution {
    public int maxSubarraySumCircular(int[] A) {
        int total = 0, maxSum = 0, max = Integer.MIN_VALUE, min = Integer.MAX_VALUE, minSum = 0;
        for (int a : A) {
            total += a;
            maxSum = Math.max(maxSum + a, a);
            minSum = Math.min(minSum + a, a);
            max = Math.max(max, maxSum);
            min = Math.min(min, minSum);
        }
        return max > 0 ? Math.max(max, total - min) : max;
    }
}

File: code/ex918.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& A) {
        int MAX = INT_MIN, MIN = INT_MAX, maxSum = 0, minSum = 0;
        int sum = accumulate(A.begin(), A.end(), 0);
        for (int a : A) {
            maxSum = max(maxSum + a, a);
            minSum = min(minSum + a, a);
            MAX = max(MAX, maxSum);
            MIN = min(MIN, minSum);
        }
        return max(MAX, MIN == sum ? INT_MIN : sum - MIN);
    }
};